main.c

```diff
@@ -379,95 +377,41 @@ inline uint8_t min(uint8_t a, uint8_t b) { return a <= b ? a : b; }
 /*
  * Update lighting table as per animation
  */
-void animationCallback(GPTDriver *_driver) {
-  (void)_driver;
+static inline void animationCallback() {
 
   // If the foreground is set we skip the animation as a way to avoid it
   // overrides the foreground
   if (!is_foregroundColor_set &&
       profiles[currentProfile].animationSpeed[currentSpeed] > 0) {
-    profiles[currentProfile].callback(ledColors, ledIntensity);
+    reactiveNeedsUpdate =
+        profiles[currentProfile].callback(ledColors, ledIntensity);
+  } else {
+    reactiveNeedsUpdate = false;
   }
 }
 
-static inline void sPWM(uint8_t cycle, uint8_t currentCount, uint8_t start,
-                        ioline_t port) {
-  if (start > 0xFF - cycle || cycle > 0xFF - start) {
-    start = 0xFF - cycle;
-  }
-  if (start <= currentCount && currentCount < start + cycle)
+static inline void sPWM(uint8_t cycle, uint8_t currentCount, ioline_t port) {
+  if (cycle > currentCount) {
+    enableRow = true;
     palSetLine(port);
-  else
+  } else {
     palClearLine(port);
+  }
 }
 
-uint16_t animationSkipCounter = 0;
-uint16_t serialCheckCounter = 0;
+uint32_t animationLastCallTime = 0;
 
-void mainCallback(GPTDriver *_driver) {
-  (void)_driver;
+uint8_t rowPWMCount = 0;
 
-  serialCheckCounter++;
-  /*
-   * If leds are disabled, this callback is called less frequently,
-   * and we should check the serial queue on every iteration.
-   * If leds are enabled, this function is called very often, and we can
-   * skip the check most of the time.
-   */
-  if (!ledEnabled || serialCheckCounter >= SERIAL_CHECK_FREQUENCY) {
-    serialCheckCounter = 0;
-    /*
-     * sdGetWouldBlock(&SD1) can be used instead which wraps iqIsEmptyI
-     * with osalSysLock/osalSysUnlock.
-     * However, this check is done very frequently, and getting the wrong
-     * data sometimes is not dangerous.
-     *    If iqIsEmptyI returns TRUE, it means there is some data to be
-     * consumed. This will never be wrong because there are no consumers of the
-     * queue other than us. It updates from TRUE to FALSE only synchronously as
-     *      a side effect of us calling `sdGet` within the loop.
-     *    If iqIsEmptyI returns FALSE, the queue might be indeed empty or it
-     * might have just received a message and we missed it. However, we will get
-     * it in the next call to columnCallback, which is safe. Therefore, the
-     * check is done without locking to speed up this function.
-     */
-    while (!iqIsEmptyI(&SD1.iqueue)) {
-      msg_t msg;
-      msg = sdGet(&SD1);
-      if (msg >= MSG_OK) {
-        executeMsg(msg);
-      }
-    }
-  }
+THD_WORKING_AREA(waThread1, 128);
```

Comment:
using a separate thread for handling messages makes message handling much more responsive. You can see it if switching profiles very quickly. Profiles get enabled/disabled quicker than before.